test/topology/util.py

```diff
@@ -7,6 +7,7 @@
 Test consistency of schema changes with topology changes.
```

Comment:
```
Because of `ring_delay_ms = 0', the generation could have been published when
it should have already been in use.
```
This is worrying.

Did you investigate which generation is the culprit? Is it one of the generations created after adding a node, or the one created by the upgrade procedure?

Even though `ring_delay_ms = 0`, new generations (apart from the first one) receive timestamps equal to `now + 2 * ring_delay + generation_leeway`, where `generation_leeway = 5s` (see `new_generation_timestamp`). IIUC, in tests, 5s should be more than enough for all nodes to learn about a new generation created after adding node. Adding a node requires full cluster availability as all topology operations do.

Hence, I suspect the generation created by the upgrade procedure. I don't know the details of the upgrade procedure, but I'm worried that if 5s isn't enough for tests, then `2 * ring_delay + generation_leeway` (usually 45 s) may not be enough in general. Well, I mean "enough in practice" because we know that committing generations is theoretically unsafe (see the large FIXME in `handle_topology_transition`, `case topology::transition_state::commit_cdc_generation`).

Also a small nit regarding this sentence: I think you meant "committed", not "published". Nodes learn about a new generation after committing it. Publishing is used only to insert data into user-facing description tables.

Reply:
> Also a small nit regarding this sentence from the commit message. I think you meant "committed", not "published". Nodes learn about a new generation after committing it. Publishing is used only to insert data into user-facing description tables.

Thanks! It's fixed now.